contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,728 @@
+     * Uses the acquired components to issue the set tokens.
+     * 
+     * @param _setToken         Address of the set token to be issued
+     * @param _minSetReceive    Minimum amount of index to receive
+     */
+    function issueSetForExactETH(
+        ISetToken _setToken,
+        uint256 _minSetReceive
+    )
+        external
+        payable
+        nonReentrant
+    {
+        
+        IWETH(WETH).deposit{value: msg.value}();
+        
+        uint256 setTokenAmount = _issueSetForExactWETH(_setToken, _minSetReceive);     // issue set token
+        
+        emit ExchangeIssue(msg.sender, address(_setToken), address(WETH), msg.value, setTokenAmount);
+        
+    }
+    
+    /**
+    * Issues an exact amount of set tokens using ERC20 tokens as input.
+    * Acquires set token components at the best price accross uniswap and sushiswap.
+    * Uses the acquired components to issue the set tokens.
+    *
+    * @param _setToken              Address of the set token to be issued
+    * @param _inputToken            Address of the input token
+    * @param _amountSetToken        Amount of set tokens to issue
+    * @param _maxAmountInputToken   Maximum amount of input tokens to be used to issue set tokens
+    */
+    function issueExactSetFromToken(
+        ISetToken _setToken,
+        IERC20 _inputToken,
+        uint256 _amountSetToken,
+        uint256 _maxAmountInputToken
+    )
+        external
+        nonReentrant
+    {
+        
+        _inputToken.transferFrom(msg.sender, address(this), _maxAmountInputToken);
+        
+        uint256 initETHAmount;
+        if(address(_inputToken) != WETH) {      // swap inputToken to WETH
+            (, Exchange exchange) = _getMaxTokenForExactToken(_maxAmountInputToken, address(_inputToken), WETH);
+            IERC20(_inputToken).approve(address(_getRouter(exchange)), _maxAmountInputToken);
+            initETHAmount = _swapExactTokensForTokens(exchange, address(_inputToken), WETH, _maxAmountInputToken);
+        } else {
+            initETHAmount = _maxAmountInputToken;
+        }
+        
+        uint256 amountETHSpent = _issueExactSetFromWETH(_setToken, _amountSetToken);        // issue set tokens
+        
+        uint256 amountETHReturn = initETHAmount.sub(amountETHSpent);        // unspent ether amount
+        
+        uint256 amountTokenReturn;        
+        if(address(_inputToken) != WETH) {      // buy return token using unspent ether
+            (, Exchange exchange) = _getMaxTokenForExactToken(amountETHReturn, WETH, address(_inputToken));
+            amountTokenReturn = _swapExactTokensForTokens(exchange, WETH, address(_inputToken), amountETHReturn);   
+        } else {
+            amountTokenReturn = amountETHReturn;
+        }
+
+        emit ExchangeIssue(msg.sender, address(_setToken), address(_inputToken), _maxAmountInputToken.sub(amountTokenReturn), _amountSetToken);
```

Comment:
Events should generally be the last thing in a function. Move the transfer above the event line